File: src/main/java/me/pcasaes/hexoids/core/domain/metrics/GameMetric.java
package me.pcasaes.hexoids.core.domain.metrics;

import pcasaes.hexoids.proto.ClientPlatforms;

import java.util.Arrays;

public class GameMetric {

    private final String name;

    private long total = 0L;

    private final long[] totalByClientPlatform;

    private GameMetric(String name) {
        this.name = name;
        totalByClientPlatform = Arrays
                .stream(ClientPlatforms.values())
                .mapToLong(a -> 0L)
                .toArray();
    }

    static GameMetric of(String name) {
        return new GameMetric(name);
    }

    public void increment(ClientPlatforms clientPlatform) {
        total++;
        if (clientPlatform == null) {
            totalByClientPlatform[ClientPlatforms.UNKNOWN.ordinal()]++;
        } else {
            totalByClientPlatform[clientPlatform.ordinal()]++;
        }
    }

    public long getTotal() {
        return total;
    }

    public long getTotalByClientPlatform(ClientPlatforms clientPlatform) {
        return totalByClientPlatform[clientPlatform.ordinal()];
    }

    public String getName() {
        return name;
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/core/domain/metrics/GameMetric.kt
package me.pcasaes.hexoids.core.domain.metrics

import pcasaes.hexoids.proto.ClientPlatforms

class GameMetric private constructor(
    private val name: String
) {


    companion object {

        fun of(name: String): GameMetric {
            return GameMetric(name)
        }
    }

    private var total = 0L

    private val totalByClientPlatform: LongArray = ClientPlatforms
        .entries
        .map { 0L }
        .toLongArray()

    fun increment(clientPlatform: ClientPlatforms?) {
        total++
        if (clientPlatform == null) {
            totalByClientPlatform[ClientPlatforms.UNKNOWN.ordinal]++
        } else {
            totalByClientPlatform[clientPlatform.ordinal]++
        }
    }

    fun getTotal(): Long {
        return total
    }

    fun getTotalByClientPlatform(clientPlatform: ClientPlatforms): Long {
        return totalByClientPlatform[clientPlatform.ordinal]
    }

    fun getName(): String {
        return name
    }

}
